osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -155,61 +156,106 @@ public async Task AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user
-                RulesetID = CurrentItem.RulesetID,
-                AllowedMods = CurrentItem.AllowedMods,
-                RequiredMods = CurrentItem.RequiredMods
-            };
-
-            newItem.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, newItem));
-            room.Playlist.Add(newItem);
-
-            await hub.OnPlaylistItemAdded(room, newItem);
+            OwnerID = CurrentItem.OwnerID,
+            BeatmapID = CurrentItem.BeatmapID,
+            BeatmapChecksum = CurrentItem.BeatmapChecksum,
+            RulesetID = CurrentItem.RulesetID,
+            AllowedMods = CurrentItem.AllowedMods,
+            RequiredMods = CurrentItem.RequiredMods
+        });
+
+        private async Task addItem(IDatabaseAccess db, MultiplayerPlaylistItem item)
+        {
+            // Add the item to the list first in order to compute gameplay order before adding to the database.
+            room.Playlist.Add(item);
+            await updatePlaylistOrder(db);
+
+            // Actually add the item to the database and invoke callback on the hub.
+            item.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, item));
+            await hub.OnPlaylistItemAdded(room, item);
         }
 
         /// <summary>
         /// Updates <see cref="CurrentItem"/> and the playlist item ID stored in the room's settings.
         /// </summary>
         private async Task updateCurrentItem()
         {
-            MultiplayerPlaylistItem newItem;
+            MultiplayerPlaylistItem nextItem = room.Playlist
+                                                   .Where(i => !i.Expired)
+                                                   .OrderBy(i => i.PlaylistOrder)
+                                                   .FirstOrDefault()
+                                               ?? room.Playlist.Last();
+
+            currentIndex = room.Playlist.IndexOf(nextItem);
+
+            long lastItemID = room.Settings.PlaylistItemId;
+            room.Settings.PlaylistItemId = nextItem.ID;
+
+            if (nextItem.ID != lastItemID)
+                await hub.OnMatchSettingsChanged(room);
+        }
+
+        /// <summary>
+        /// Updates the order of items in the playlist according to the queueing mode.
+        /// </summary>
+        private async Task updatePlaylistOrder(IDatabaseAccess db)
+        {
+            List<MultiplayerPlaylistItem> orderedActiveItems;
 
             switch (room.Settings.QueueMode)
             {
                 default:
-                    // Pick the first available non-expired playlist item, or default to the last item for when all items are expired.
-                    newItem = room.Playlist.FirstOrDefault(i => !i.Expired) ?? room.Playlist.Last();
+                    orderedActiveItems = room.Playlist.Where(item => !item.Expired).OrderBy(item => item.ID == 0 ? int.MaxValue : item.ID).ToList();
```

Comment:
So I presume the 0 special case is to accommodate the `addItem()` usage? Because in that case the model is sorted before being saved to DB and therefore being assigned an ID?

If that is correct I'd probably leave an inline comment here because at face value I find that behaviour kinda weird...

Reply:
I've cleaned this up a bit by setting `long.MaxValue` in `addItem`, which removes these checks scattered across this method. The only remaining one is at the very bottom of this method which previously had and still does have an inline comment.